test/Microsoft.Health.Fhir.Tests.E2E/Rest/Audit/AuditTests.cs

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Rest;
+using Microsoft.Health.Fhir.Api.Features.Audit;
+using Microsoft.Health.Fhir.Tests.Common;
+using Microsoft.Health.Fhir.Tests.E2E.Common;
+using Xunit;
+using FhirClient = Microsoft.Health.Fhir.Tests.E2E.Common.FhirClient;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Tests.E2E.Rest.Audit
+{
+    public class AuditTests : IClassFixture<AuditTestFixture>
+    {
+        private const string RequestIdHeaderName = "X-Request-Id";
+
+        private readonly AuditTestFixture _fixture;
+        private readonly FhirClient _client;
+
+        private readonly TraceAuditLogger _auditLogger;
+
+        public AuditTests(AuditTestFixture fixture)
+        {
+            _fixture = fixture;
+            _client = fixture.FhirClient;
+            _auditLogger = _fixture.AuditLogger;
+        }
+
+        [Fact]
+        public async Task GivenAResource_WhenCreated_ThenAuditLogEntriesShouldBeCreated()
+        {
+            await ExecuteAndValidate(
+                () => _client.CreateAsync(Samples.GetDefaultObservation()),
+                "create",
+                ResourceType.Observation,
+                _ => "Observation",
+                HttpStatusCode.Created);
+        }
+
+        [Fact]
+        public async Task GivenAnExistingResource_WhenRead_ThenAuditLogEntriesShouldBeCreated()
+        {
+            FhirResponse<Patient> result = null;
+
+            await ExecuteAndValidate(
+                async () =>
+                {
+                    result = await _client.CreateAsync(Samples.GetDefaultPatient());
+
+                    return await _client.ReadAsync<Patient>(ResourceType.Patient, result.Resource.Id);
+                },
+                "read",
+                ResourceType.Patient,
+                p => $"Patient/{p.Id}",
+                HttpStatusCode.OK);
+        }
+
+        [Fact]
+        public async Task GivenANonExistingResource_WhenRead_ThenAuditLogEntriesShouldBeCreated()
+        {
+            await ExecuteAndValidate(
+                async () =>
+                {
+                    FhirResponse<OperationOutcome> result = null;
+
+                    try
+                    {
+                        await _client.ReadAsync<Patient>(ResourceType.Patient, "123");
+                    }
+                    catch (FhirException ex)
+                    {
+                        result = ex.Response;
+                    }
+
+                    return result;
+                },
+                "read",
+                ResourceType.OperationOutcome,
```

Comment:
There are some issues with the existing audit event logging logic. For example, when a non-existing resource is read, it logs the resource type as OperationOutcome, which is incorrect. In the delete test below, it logs the resource type as null, which I think is incorrect as well. In both cases, we should log the actual resource type where the operation is being executed on, which is Patient in these two cases. I will file a separate issue to fix that.

Reply:
Filed issue and added comments.